### PR TITLE
[IMP] sms: sms composer promenade

### PR DESCRIPTION
**Pull request by mish-odoo**
Current behavior before PR:

There is an extra space at the top of the SMS composer and
an extra space below the validation message.
The validation message has no model description.

Desired behavior after PR is merged:

Below points:

1/ Remove additional space on top of the SMS composer.
2/ Remove the additional space below the validation message.
3/ Improve the validation message of the invalid phone number.

This is the goal of this commit.

PR #83062
Task-2710478

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sms/wizard/sms_composer.py [hunk 1]
```diff
@@ -1,6 +1,5 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
-
```
Comment (awa-odoo) [inline addons/sms/wizard/sms_composer.py:3]: Unnecessary change.

File: addons/sms/wizard/sms_composer_views.xml [hunk 2]
```diff
@@ -21,13 +21,13 @@
                         <field name="sanitized_numbers" invisible="1"/>
 
                         <!-- Single mode information (invalid number) -->
-                        <div colspan="2" class="alert alert-danger text-center mb-3" role="alert"
+                        <div colspan="2" class="alert alert-danger text-center mb-3 mt-n4" role="alert"
```
Comment (awa-odoo) [inline addons/sms/wizard/sms_composer_views.xml:24]: Have you tried moving that warning the the header section of the form?
I'm not a fan of using negative margin, seems very hackish.

File: addons/sms/wizard/sms_composer_views.xml [hunk 3]
```diff
@@ -21,13 +21,13 @@
...
                             attrs="{'invisible': ['|', ('comment_single_recipient', '=', False), ('recipient_single_valid', '=', True)]}">
-                            <p class="my-0">Invalid phone number</p>
+                            <p class="my-0">Invalid number, make sure to set a country on the <field name="res_description" readonly="1"/> or to specify the country code.</p>
                         </div>
 
                         <!-- Mass mode information (res_ids versus active domain) -->
-                        <div colspan="2" class="alert alert-info text-center mb-3" role="alert"
+                        <div colspan="2" class="alert alert-info text-center mb-3 mt-n4" role="alert"
```
Comment (awa-odoo) [inline addons/sms/wizard/sms_composer_views.xml:30]: Same remark here.

File: addons/sms/wizard/sms_composer.py [hunk 4]
```diff
@@ -37,6 +36,7 @@ def default_get(self, fields):
         ('mass', 'Send SMS in batch')], string='Composition Mode',
         compute='_compute_composition_mode', precompute=True, readonly=False, required=True, store=True)
     res_model = fields.Char('Document Model Name')
+    res_description = fields.Char('Document Model Description', compute='_compute_res_description')
```
Comment (awa-odoo) [inline addons/sms/wizard/sms_composer.py:39]: Let's rename to "res_model_description".

**Review by awa-odoo — COMMENTED** [review]
Hello @mish-odoo 

Added some light comments to try to clean as much as possible.

About the message flicker point, please refer to the task chatter, I think it's already been fixed elsewhere.

Thanks & cheers!

File: addons/sms/wizard/sms_composer_views.xml (hunk 2 above)
Comment (mish-odoo) [reply addons/sms/wizard/sms_composer_views.xml:24]: Hello @awa-odoo 

I tried with the header section but the header is not available in the new window. 
Please give suggestions for a negative margin.

Thanks!

Comment (awa-odoo) [reply addons/sms/wizard/sms_composer_views.xml:24]: Hello @mish-odoo 

Then you can put those warnings just after the `<sheet>` tag, outside of any `<group>`.
It looks very slightly different from your version but it's better technically speaking.

(Also, you can add a space at the end of the 2 first spans, otherwise it looks like this:
"Invalid number, make sure to set a country on theContactorto specify the country code.")

Cheers!

File: addons/sms/wizard/sms_composer.py [hunk 5]
```diff
@@ -83,6 +84,14 @@ def _compute_composition_mode(self):
                 else:
                     composer.composition_mode = 'comment'
 
+    @api.depends('res_model')
+    def _compute_res_model_description(self):
+        for composer in self:
```
Comment (tde-banana-odoo) [inline addons/sms/wizard/sms_composer.py:89]: Maybe we could batch the easiest part, like
```
self.res_model_description = False
for composer in self.filtered('res_model'):
    composer.res_model_description = self.env['ir.model']._get(composer.res_model).display_name
```

File: addons/sms/wizard/sms_composer.py [hunk 6]
```diff
@@ -139,8 +147,7 @@ def _compute_recipient_single_valid(self):
         for composer in self:
             value = composer.recipient_single_number_itf or composer.recipient_single_number
             if value:
-                records = composer._get_records()
-                sanitized = phone_validation.phone_sanitize_numbers_w_record([value], records)[value]['sanitized']
+                sanitized = phone_validation.phone_sanitize_numbers_w_record([value], composer)[value]['sanitized']
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/sms/wizard/sms_composer.py:151]: Up ^^

**Review by tde-banana-odoo — COMMENTED** [review]
Pouic validation

File: addons/sms/wizard/sms_composer.py (hunk 5 above)
Comment (nla-odoo) [reply addons/sms/wizard/sms_composer.py:89]: Hello @tde-banana-odoo 

if we don't have res_model then `res_model_description` will be  unassigned and it will raise the error.
for better we can use inline condition.

Thanks

Comment (tde-banana-odoo) [reply addons/sms/wizard/sms_composer.py:89]: Hence the first part: first assign False (self.res_model_description = False), then update only relevant :) .

File: addons/sms/wizard/sms_composer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nla-odoo) [reply addons/sms/wizard/sms_composer.py:151]: Before we have Country in the composer and we need the latest value of country, so we did this.
Indeed this change was out of focus.

Thanks

File: addons/sms/wizard/sms_composer_views.xml [hunk 7]
```diff
@@ -5,6 +5,24 @@
         <field name="model">sms.composer</field>
         <field name="arch" type="xml">
             <form string="Send an SMS">
+                <!-- Single mode information (invalid number) -->
+                <div colspan="2" class="alert alert-danger text-center mb-3" role="alert"
+                    attrs="{'invisible': ['|', ('comment_single_recipient', '=', False), ('recipient_single_valid', '=', True)]}">
+                    <p class="my-0">
+                        <span>Invalid number, make sure to set a country on the </span>
+                        <span><field name="res_model_description" readonly="1"/></span>
```
Comment (awa-odoo) [inline addons/sms/wizard/sms_composer_views.xml:13]: Since "res_model_description" can be void according to field compute method, shouldn't we handle the case where it's not set?

We don't want the message to be:

> Invalid number, make sure to set a country on the False or to specify the country code.

Comment (awa-odoo) [inline addons/sms/wizard/sms_composer_views.xml:13]: Also, don't think readonly needs to be specified as this is a regular compute field.

File: addons/sms/static/src/js/fields_sms_widget.js [hunk 8]
```diff
@@ -30,26 +30,14 @@ var SmsWidget = FieldTextEmojis.extend({
...
-        var def = this._super.apply(this, arguments);
-
-        this._compute();
-        var $sms_container = $('<div class="o_sms_container"/>');
-        $sms_container.append(this._renderSMSInfo());
-        $sms_container.append(this._renderIAPButton());
-        this.$el = this.$el.add($sms_container);
-
-        return def;
+        if (this.mode === 'edit') {
```
Comment (awa-odoo) [inline addons/sms/static/src/js/fields_sms_widget.js:33]: Do we need all those changes just to remove the extra underline?
Seems unlikely, looks like you just re-used your previous code.
(The commit message even mentions the focus issue while it should be fixed already).

Please investigate only the underline issue and make a fix with the minimum amount of code changes.

**Review by awa-odoo — COMMENTED** [review]
Hello @mish-odoo 

Some extra comments, I think we can reduce the amount of changes here.
Please refer to details in each comment.

Thanks & cheers!

File: addons/sms/wizard/sms_composer_views.xml (hunk 7 above)
Comment (mish-odoo) [reply addons/sms/wizard/sms_composer_views.xml:13]: Hello @awa-odoo 

'False' is not set in the res_model_description. display_name is set based on the model name.

Thanks!

**Comment by awa-odoo** [comment]
@mish-odoo LGTM.

Only a small typo left in the commit message:

> Imporve

-> 

> Improve
